isobus/include/isobus/isobus/can_parameter_group_builder.hpp

```diff
@@ -0,0 +1,503 @@
+				while (bits);
+				// No additional masking required in this case.
+				return true;
+			}
+
+			// ----------------------------------------
+			//  Stage one - fill out the current byte.
+			// ----------------------------------------
+
+			// All this data can come from the first byte of the input.
+			// Everything will fit in the current byte, which must mean there's at most one byte
+			// of data to write.  Hence we put this version first because it covers more single
+			// byte cases.
+			std::uint8_t mask = (1 << offset) - 1;
+			buffer[byte] = (buffer[byte] & mask) | (*data << offset);
+			bits -= remaining;
+			++byte;
+
+			// -------------------------------
+			//  Stage two - copy whole bytes.
+			// -------------------------------
+
+			while (bits > 8)
+			{
+				buffer[byte] = (*data >> remaining);
+				++data;
+				buffer[byte] = (buffer[byte] & mask) | (*data << offset);
+				bits -= 8;
+				++byte;
+			}
+
+			// -------------------------------
+			//  Stage three - copy whole bytes.
+			// -------------------------------
+
+			// `bits` is number of bits left.  It may or may not span a byte boundary in the input.
+			if (bits == 0)
+			{
+				// Nothing left to copy.
+				if (writeOffset % 8 != 0)
+				{
+					// Mask the top bits to hide excess data.
+					buffer[byte] |= 255 << (writeOffset % 8);
+				}
+				return true;
+			}
+
+			// I'm sure this code can be compressed with the code above.
+			buffer[byte] = (*data >> remaining);
+			if (offset < bits)
+			{
+				// The final output spans two bytes of input.
+				++data;
+				buffer[byte] = (buffer[byte] & mask) | (*data << offset);
+			}
+			if (writeOffset % 8 != 0)
+			{
+				// Mask the top bits to hide excess data.
+				buffer[byte] |= 255 << (writeOffset % 8);
+			}
+
+			return true;
+		}
+		
+		bool read_bits(std::uint8_t * data, std::size_t bits)
+		{
+			std::size_t byte = get_read_byte_offset();
+			std::size_t input = get_read_bit_offset();
+			std::size_t remaining = 8 - input;
+			std::size_t output = 0;
+			std::size_t space = 8 - output;
+			// Mark as read, even though we actually haven't yet.
+			readOffset += bits;
+			if (readOffset > writeOffset)
+			{
+				// Trying to read too much data.
+				readOffset -= bits;
+				return false;
+			}
+			// Initialise the current destination byte.
+			*data = 0;
+			while (bits)
+			{
+				// Move in the new data.
+				*data = (*data & (255 >> space)) | ((buffer[byte] >> input) << output);
+				if (space >= bits)
+				{
+					space -= bits;
+					break;
+				}
+				if (space > remaining)
+				{
+					space -= remaining;
+					output = 8 - space;
+					remaining = 0;
+					input = 0;
+					++byte;
+				}
+				else if (remaining > space)
+				{
+					bits -= space;
+					remaining -= space;
+					input = 8 - remaining;
+					space = 8;
+					output = 0;
+					++data;
+					*data = 0;
+				}
+				else
+				{
+					bits -= space;
+					// Perfectly matched.
+					remaining = 8;
+					input = 0;
+					space = 8;
```

Comment:
You use the magic number 8 frequently. If it's just the number of bits in a byte, it might be nicer to get that constant from `#include <limits>` or otherwise define a `static constexpr std::uint8_t` for it